hikari/impl/interaction_server.py

```diff
@@ -0,0 +1,558 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Standard implementation of a REST based interactions server."""
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = ["InteractionServer"]
+
+import asyncio
+import logging
+import sys
+import typing
+
+import aiohttp.web
+import aiohttp.web_runner
+
+from hikari import errors
+from hikari.api import interaction_server
+from hikari.api import special_endpoints
+from hikari.interactions import bases
+from hikari.internal import data_binding
+from hikari.internal import ed25519
+from hikari.internal import ux
+
+if typing.TYPE_CHECKING:
+    import socket as socket_
+    import ssl
+
+    import aiohttp.typedefs
+
+    from hikari import applications
+    from hikari.api import entity_factory as entity_factory_
+    from hikari.api import event_factory as event_factory_
+    from hikari.api import rest as rest_client_
+
+    ListenerDictT = typing.Dict[
+        typing.Type[interaction_server.InteractionT],
+        interaction_server.ListenerT[interaction_server.InteractionT, interaction_server.ResponseT],
+    ]
+
+
+_LOGGER: typing.Final[logging.Logger] = logging.getLogger("hikari.interaction_server")
+
+# Internal interaction and interaction response types.
+_PING_INTERACTION_TYPE: typing.Final[int] = 1
+_PONG_RESPONSE_TYPE: typing.Final[int] = 1
+
+# HTTP status codes.
+_OK_STATUS: typing.Final[int] = 200
+_BAD_REQUEST_STATUS: typing.Final[int] = 400
+_PAYLOAD_TOO_LARGE_STATUS: typing.Final[int] = 413
+_UNSUPPORTED_MEDIA_TYPE_STATUS: typing.Final[int] = 415
+_INTERNAL_SERVER_ERROR_STATUS: typing.Final[int] = 500
+_NOT_IMPLEMENTED: typing.Final[int] = 501
+
+# Header keys and values
+_X_SIGNATURE_ED25519_HEADER: typing.Final[str] = "X-Signature-Ed25519"
+_X_SIGNATURE_TIMESTAMP_HEADER: typing.Final[str] = "X-Signature-Timestamp"
+_CONTENT_TYPE_KEY: typing.Final[str] = "Content-Type"
+_USER_AGENT_KEY: typing.Final[str] = "User-Agent"
+_JSON_CONTENT_TYPE: typing.Final[str] = "application/json"
+_TEXT_CONTENT_TYPE: typing.Final[str] = "text/plain"
+
+# Constant response payloads
+_PONG_PAYLOAD: typing.Final[str] = data_binding.dump_json({"type": _PONG_RESPONSE_TYPE})
```

Comment:
That's not where interning would benefit anything:
```
Enter string in the table of “interned” strings and return the interned string – which is string itself or a copy. Interning strings is useful to gain a little performance on dictionary lookup – if the keys in a dictionary are interned, and the lookup key is interned, the key comparisons (after hashing) can be done by a pointer compare instead of a string compare. Normally, the names used in Python programs are automatically interned, and the dictionaries used to hold module, class or instance attributes have interned keys.
```
and strings are implicitly interned under cpython anyway